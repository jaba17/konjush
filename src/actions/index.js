export const SETTING_TOGGLE_VISIBILITY    = 'SETTING_TOGGLE_VISIBILITY';
export const SETTING_ORDER                = 'SETTING_ORDER';
export const CARD_VIEW_NEGATIVE           = 'CARD_VIEW_NEGATIVE';
export const CARD_VIEW_POSITIVE           = 'CARD_VIEW_POSITIVE';
export const RECEIVE_CONJUGATIONS         = 'RECEIVE_CONJUGATIONS';
export const RECEIVE_URL                  = 'RECEIVE_URL';
export const UPDATE_INPUT                 = 'UPDATE_INPUT';