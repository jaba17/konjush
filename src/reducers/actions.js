export const SETTING_TOGGLE_VISIBILITY                = 'SETTING_TOGGLE_VISIBILITY';
export const SETTING_ORDER                            = 'SETTING_ORDER';
export const CARD_VIEW_NEGATIVE                       = 'CARD_VIEW_NEGATIVE';
export const CARD_VIEW_POSITIVE                       = 'CARD_VIEW_POSITIVE';
export const RECEIVE_CONJUGATIONS                     = 'RECEIVE_CONJUGATIONS';
export const RECEIVE_URL                              = 'RECEIVE_URL';
export const ADD_RECENT_VERB                          = 'ADD_RECENT_VERB';
export const RECEIVE_CONJUGATIONS_URL_AND_RECENT_VERB = 'RECEIVE_CONJUGATIONS_URL_AND_RECENT_VERB';