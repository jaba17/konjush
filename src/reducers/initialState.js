let initialState = {
  conjugator: {
    url: '/konjush/conjugation',
    conjugations: false,
    cardStates: [true, true, true, true, true, true, true, true, true, true],
    recentVerbs: []
  },
  settings: [
    {
      title: 'Aorist',
      tense: 'aorist',
      visible: true
    },
    {
      title: 'Gerund',
      tense: 'gerund',
      visible: true
    },
    {
      title: 'Future',
      tense: 'future',
      visible: true
    },
    {
      title: 'Simple Past',
      tense: 'simple-past',
      visible: true
    },
    {
      title: 'Mi≈ü Past',
      tense: 'mis-past',
      visible: true
    },
    {
      title: 'Past Continuous',
      tense: 'past-continuous',
      visible: true
    },
    {
      title: 'Present Indefinite',
      tense: 'present-indefinite',
      visible: true
    },
    {
      title: 'Gerund Indefinite',
      tense: 'present-continuous-indefinite',
      visible: true
    },
    {
      title: 'Future Indefinite',
      tense: 'future-indefinite',
      visible: true
    },
    {
      title: 'Potential',
      tense: 'potential',
      visible: true
    },
  ],
}

export default initialState;