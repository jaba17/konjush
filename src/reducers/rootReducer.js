import conjugator from './conjugatorReducer';
import settings from './settingsReducer';

export {
  conjugator,
  settings
}