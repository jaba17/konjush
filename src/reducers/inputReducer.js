import { UPDATE_INPUT } from './../actions';
export default (state = null, action) => action.type === UPDATE_INPUT ? action.input : state;