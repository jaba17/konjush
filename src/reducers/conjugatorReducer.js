import { 
  CARD_VIEW_POSITIVE, 
  CARD_VIEW_NEGATIVE, 
  RECEIVE_CONJUGATIONS, 
  RECEIVE_URL, 
  ADD_RECENT_VERB,
  RECEIVE_CONJUGATIONS_URL_AND_RECENT_VERB
} from './actions';

import initialState from './initialState';

const appState = (JSON.parse(window.localStorage.getItem('store')) || initialState).conjugator;

const conjugator = (state = appState, action) => {
  switch(action.type){
    case CARD_VIEW_NEGATIVE:
      return {
        ...state,
        cardStates: state.cardStates.map((cardState, index) => index === action.at ? { positive: false } : cardState)
      }
    case CARD_VIEW_POSITIVE:
      return {
        ...state,
        cardStates: state.cardStates.map((cardState, index) => index === action.at ? { positive: true } : cardState)
      }
    case RECEIVE_CONJUGATIONS:
      return {
        ...state,
        conjugations: action.conjugations
      }
    case RECEIVE_URL:
      return {
        ...state,
        url: action.url
      }
    case ADD_RECENT_VERB:
      return {
        ...state,
        recentVerbs: addRecentVerb(state, action)
      }
    case RECEIVE_CONJUGATIONS_URL_AND_RECENT_VERB:
      return {
        ...state,
        conjugations: action.conjugations,
        url: action.url,
        recentVerbs: addRecentVerb(state, action)
      }
    default:
      return state;
  }
}

const addRecentVerb = (state, action) => {
  return [
    action.verb, 
    ...(state.recentVerbs.includes(action.verb) ? state.recentVerbs.filter(e => e !== action.verb) : state.recentVerbs)
  ].slice(0, 100)
}

export default conjugator;
