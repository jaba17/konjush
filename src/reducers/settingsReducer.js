import { SETTING_TOGGLE_VISIBILITY, SETTING_ORDER } from './actions';
import initialState from './initialState';
import { map } from './../utils/arrays';

const appState = (JSON.parse(window.localStorage.getItem('store')) || initialState).settings;

const settings = (state = appState, action) => {

  switch(action.type){
    case SETTING_TOGGLE_VISIBILITY:
      return map(state, setting => action.tense === setting.tense ? { ...setting, visible: !setting.visible } : setting);
    case SETTING_ORDER:
      return reorderArray(state, action.from, action.to);
    default:
      return state;
  }

}

function reorderArray(arr, position, to){
  arr = arr.slice();
  let target = arr.splice(position, 1)[0];
  return [...arr.slice(0, to), target, ...arr.slice(to)];
}

export default settings;