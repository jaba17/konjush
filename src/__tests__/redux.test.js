import conjugatorReducer from './../reducers/conjugatorReducer';
import settingsReducer from './../reducers/settingsReducer';
import initialState from './../reducers/initialState';
import { SETTING_TOGGLE_VISIBILITY, SETTING_ORDER, RECEIVE_URL_AND_RECENT_VERB, RECEIVE_URL, CARD_TOGGLE_VIEW_STATE } from './../reducers/actions';


describe('>>> conjugatorReducer <<<', () => {

  it('action CARD_TOGGLE_VIEW_STATE', () => {

    expect(
      conjugatorReducer(initialState.conjugator, {
        type: CARD_TOGGLE_VIEW_STATE,
        at: 0,
        view: false
      })
    ).toMatchSnapshot();

    expect(
      conjugatorReducer(initialState.conjugator, {
        type: CARD_TOGGLE_VIEW_STATE,
        at: 4,
        view: false
      })
    ).toMatchSnapshot();

  })

  it('action RECEIVE_URL', () => {

    expect(
      conjugatorReducer(initialState.conjugator, {
        type: RECEIVE_URL,
        url: '/konjush/conjugation/yemek'
      })
    ).toMatchSnapshot()

    expect(
      conjugatorReducer(initialState.conjugator, {
        type: RECEIVE_URL,
        url: '/konjush/'
      })
    ).toMatchSnapshot()

  })

  it('action RECEIVE_URL_AND_RECENT_VERB', () => {

    expect(
      conjugatorReducer(initialState.conjugator, {
        type: RECEIVE_URL_AND_RECENT_VERB,
        url: '/konjush/conjugation/olmak',
        verb: 'olmak'
      })
    ).toMatchSnapshot()
  })

})

describe('>>> settingsReducers <<<', () => {

  it('action SETTING_TOGGLE_VISIBILITY', () => {

    expect(
      settingsReducer(initialState.settings, {
        type: SETTING_TOGGLE_VISIBILITY,
        tense: 'simple-past',
      })
    ).toMatchSnapshot()

  })

  it('action SETTING_ORDER', () => {

    expect(
      settingsReducer(initialState.settings, {
        type: SETTING_ORDER,
        from: 1,
        to: 0
      })
    ).toMatchSnapshot()
    
  })

})