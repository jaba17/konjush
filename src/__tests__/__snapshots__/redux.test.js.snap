// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> conjugatorReducer <<< action CARD_TOGGLE_VIEW_STATE 1`] = `
Object {
  "cardStates": Array [
    false,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
  ],
  "conjugations": false,
  "recentVerbs": Array [],
  "url": "/konjush/conjugation",
}
`;

exports[`>>> conjugatorReducer <<< action CARD_TOGGLE_VIEW_STATE 2`] = `
Object {
  "cardStates": Array [
    true,
    true,
    true,
    true,
    false,
    true,
    true,
    true,
    true,
    true,
  ],
  "conjugations": false,
  "recentVerbs": Array [],
  "url": "/konjush/conjugation",
}
`;

exports[`>>> conjugatorReducer <<< action RECEIVE_URL 1`] = `
Object {
  "cardStates": Array [
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
  ],
  "conjugations": false,
  "recentVerbs": Array [],
  "url": "/konjush/conjugation/yemek",
}
`;

exports[`>>> conjugatorReducer <<< action RECEIVE_URL 2`] = `
Object {
  "cardStates": Array [
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
  ],
  "conjugations": false,
  "recentVerbs": Array [],
  "url": "/konjush/",
}
`;

exports[`>>> conjugatorReducer <<< action RECEIVE_URL_AND_RECENT_VERB 1`] = `
Object {
  "cardStates": Array [
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
    true,
  ],
  "conjugations": false,
  "recentVerbs": Array [
    "olmak",
  ],
  "url": "/konjush/conjugation/olmak",
}
`;

exports[`>>> settingsReducers <<< action SETTING_ORDER 1`] = `
Array [
  Object {
    "tense": "gerund",
    "title": "Gerund",
    "visible": true,
  },
  Object {
    "tense": "aorist",
    "title": "Aorist",
    "visible": true,
  },
  Object {
    "tense": "future",
    "title": "Future",
    "visible": true,
  },
  Object {
    "tense": "simple-past",
    "title": "Simple Past",
    "visible": true,
  },
  Object {
    "tense": "mis-past",
    "title": "Miş Past",
    "visible": true,
  },
  Object {
    "tense": "past-continuous",
    "title": "Past Continuous",
    "visible": true,
  },
  Object {
    "tense": "present-indefinite",
    "title": "Present Indefinite",
    "visible": true,
  },
  Object {
    "tense": "present-continuous-indefinite",
    "title": "Gerund Indefinite",
    "visible": true,
  },
  Object {
    "tense": "future-indefinite",
    "title": "Future Indefinite",
    "visible": true,
  },
  Object {
    "tense": "potential",
    "title": "Potential",
    "visible": true,
  },
]
`;

exports[`>>> settingsReducers <<< action SETTING_TOGGLE_VISIBILITY 1`] = `
Array [
  Object {
    "tense": "aorist",
    "title": "Aorist",
    "visible": true,
  },
  Object {
    "tense": "gerund",
    "title": "Gerund",
    "visible": true,
  },
  Object {
    "tense": "future",
    "title": "Future",
    "visible": true,
  },
  Object {
    "tense": "simple-past",
    "title": "Simple Past",
    "visible": false,
  },
  Object {
    "tense": "mis-past",
    "title": "Miş Past",
    "visible": true,
  },
  Object {
    "tense": "past-continuous",
    "title": "Past Continuous",
    "visible": true,
  },
  Object {
    "tense": "present-indefinite",
    "title": "Present Indefinite",
    "visible": true,
  },
  Object {
    "tense": "present-continuous-indefinite",
    "title": "Gerund Indefinite",
    "visible": true,
  },
  Object {
    "tense": "future-indefinite",
    "title": "Future Indefinite",
    "visible": true,
  },
  Object {
    "tense": "potential",
    "title": "Potential",
    "visible": true,
  },
]
`;
